File: scripts/extract-frames.js
#!/usr/bin/env node
/**
 * Frame Extraction Script for Convex Vision Analysis (Node.js)
 * 
 * Extracts frames from video files and sends them to Convex API
 * 
 * Requirements:
 *   npm install fluent-ffmpeg axios
 * 
 * Usage:
 *   node extract-frames.js --video path/to/video.mp4 --camera-id <id> --api-url <url>
 *   node extract-frames.js --youtube <url> --camera-id <id> --api-url <url>
 */

const ffmpeg = require('fluent-ffmpeg');
const axios = require('axios');
const fs = require('fs');
const path = require('path');

class FrameExtractor {
  constructor(apiUrl, cameraId, targetFps = 0.5, priority = 1) {
    this.apiUrl = apiUrl.replace(/\/$/, '') + '/api/analyze-frame';
    this.cameraId = cameraId;
    this.targetFps = targetFps;
    this.priority = priority;
    this.framesSent = 0;
    this.framesCached = 0;
    this.framesSkipped = 0;
  }

  async extractFromVideo(videoPath) {
    if (!fs.existsSync(videoPath)) {
      throw new Error(`Video file not found: ${videoPath}`);
    }

    console.log(`📹 Processing video: ${videoPath}`);

    // Create temp directory for frames
    const tempDir = path.join(__dirname, 'temp_frames');
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir);
    }

    return new Promise((resolve, reject) => {
      let frameCount = 0;

      ffmpeg(videoPath)
        .fps(this.targetFps)
        .on('start', (cmd) => {
          console.log(`🎬 Starting extraction at ${this.targetFps} FPS`);
        })
        .on('progress', (progress) => {
          if (progress.frames) {
            process.stdout.write(`\r📊 Extracting... ${progress.frames} frames`);
          }
        })
        .on('end', async () => {
          console.log('\n✅ Extraction complete! Sending to API...');
          
          // Send all extracted frames
          const files = fs.readdirSync(tempDir);
          for (const file of files.sort()) {
            if (file.endsWith('.jpg')) {
              await this.sendFrame(path.join(tempDir, file), frameCount++);
            }
          }

          // Cleanup
          fs.rmSync(tempDir, { recursive: true });

          console.log('\n📈 Summary:');
          console.log(`   Sent: ${this.framesSent}`);
          console.log(`   Cached: ${this.framesCached}`);
          console.log(`   Skipped: ${this.framesSkipped}`);
          
          resolve();
        })
        .on('error', (err) => {
          console.error('❌ Error:', err.message);
          reject(err);
        })
        .save(path.join(tempDir, 'frame-%04d.jpg'));
    });
  }

  async sendFrame(framePath, frameNumber) {
    // Read and encode frame
    const frameBuffer = fs.readFileSync(framePath);
    const frameB64 = frameBuffer.toString('base64');

    try {
      const response = await axios.post(this.apiUrl, {
        cameraId: this.cameraId,
        frameData: frameB64,
        priority: this.priority,
      }, {
        timeout: 30000,
      });

      const result = response.data;

      // Track results
      let status;
      if (result.cached) {
        this.framesCached++;
        status = '💾 CACHED';
      } else if (result.skipped) {
        this.framesSkipped++;
        status = '⏭️  SKIPPED';
      } else if (result.queued) {
        this.framesSent++;
        status = `📤 QUEUED ${result.message || ''}`;
      } else {
        status = '❓ UNKNOWN';
      }

      console.log(`Frame ${frameNumber}: ${status}`);
      
    } catch (error) {
      console.error(`❌ Error sending frame ${frameNumber}:`, error.message);
    }
  }

  async extractFromYouTube(youtubeUrl) {
    console.log(`📺 Downloading from YouTube: ${youtubeUrl}`);
    
    const tempVideo = path.join(__dirname, 'temp_youtube.mp4');
    
    // Use youtube-dl or yt-dlp to download
    const { exec } = require('child_process');
    
    return new Promise((resolve, reject) => {
      exec(`yt-dlp -f worst -o "${tempVideo}" "${youtubeUrl}"`, async (error) => {
        if (error) {
          reject(new Error('Failed to download YouTube video. Make sure yt-dlp is installed.'));
          return;
        }

        console.log('✅ Download complete!');
        
        // Extract frames
        await this.extractFromVideo(tempVideo);
        
        // Cleanup
        if (fs.existsSync(tempVideo)) {
          fs.unlinkSync(tempVideo);
        }
        
        resolve();
      });
    });
  }
}

// CLI
if (require.main === module) {
  const args = process.argv.slice(2);
  
  const getArg = (flag) => {
    const index = args.indexOf(flag);
    return index !== -1 && args[index + 1] ? args[index + 1] : null;
  };

  const hasFlag = (flag) => args.includes(flag);

  const video = getArg('--video');
  const youtube = getArg('--youtube');
  const cameraId = getArg('--camera-id');
  const apiUrl = getArg('--api-url');
  const fps = parseFloat(getArg('--fps') || '0.5');
  const priority = parseInt(getArg('--priority') || '1');

  if (!cameraId || !apiUrl) {
    console.error('❌ Missing required arguments: --camera-id and --api-url');
    console.log('\nUsage:');
    console.log('  node extract-frames.js --video <path> --camera-id <id> --api-url <url>');
    console.log('  node extract-frames.js --youtube <url> --camera-id <id> --api-url <url>');
    console.log('\nOptions:');
    console.log('  --fps <number>      Frames per second (default: 0.5)');
    console.log('  --priority <1-10>   Priority level (default: 1)');
    process.exit(1);
  }

  if (!video && !youtube) {
    console.error('❌ Must specify either --video or --youtube');
    process.exit(1);
  }

  const extractor = new FrameExtractor(apiUrl, cameraId, fps, priority);

  (async () => {
    try {
      if (video) {
        await extractor.extractFromVideo(video);
      } else if (youtube) {
        await extractor.extractFromYouTube(youtube);
      }
      console.log('✅ Done!');
    } catch (error) {
      console.error('❌ Error:', error.message);
      process.exit(1);
    }
  })();
}

module.exports = { FrameExtractor };

File: scripts/extract-frames.py
#!/usr/bin/env python3
"""
Frame Extraction Script for Convex Vision Analysis

This script extracts frames from video files or YouTube videos and sends them
to the Convex API for analysis with intelligent batching.

Requirements:
    pip install opencv-python requests yt-dlp pillow

Usage:
    # From local video file
    python extract-frames.py --video path/to/video.mp4 --camera-id <convex_camera_id> --api-url <your_convex_url>
    
    # From YouTube URL
    python extract-frames.py --youtube https://youtube.com/watch?v=... --camera-id <camera_id> --api-url <your_api_url>
    
    # With custom frame rate
    python extract-frames.py --video video.mp4 --camera-id <id> --api-url <url> --fps 1
"""

import cv2
import base64
import requests
import argparse
import time
import os
from typing import Optional
from pathlib import Path

class FrameExtractor:
    def __init__(self, api_url: str, camera_id: str, target_fps: float = 0.5, priority: int = 1):
        """
        Initialize frame extractor
        
        Args:
            api_url: Full Convex deployment URL (e.g., https://accurate-marlin-326.convex.site)
            camera_id: Convex camera feed ID
            target_fps: Frames per second to extract (default: 0.5 = 1 frame every 2 seconds)
            priority: Priority for frame processing (1-10, higher = more urgent)
        """
        self.api_url = api_url.rstrip('/') + '/api/analyze-frame'
        self.camera_id = camera_id
        self.target_fps = target_fps
        self.priority = priority
        self.frames_sent = 0
        self.frames_cached = 0
        self.frames_skipped = 0
        
    def extract_from_video(self, video_path: str) -> None:
        """Extract frames from a local video file"""
        if not os.path.exists(video_path):
            raise FileNotFoundError(f"Video file not found: {video_path}")
        
        print(f"📹 Opening video: {video_path}")
        cap = cv2.VideoCapture(video_path)
        
        if not cap.isOpened():
            raise ValueError("Failed to open video file")
        
        # Get video properties
        video_fps = cap.get(cv2.CAP_PROP_FPS)
        total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
        duration = total_frames / video_fps
        
        print(f"📊 Video info: {video_fps:.2f} FPS, {total_frames} frames, {duration:.2f}s duration")
        print(f"🎯 Extracting at {self.target_fps} FPS")
        
        # Calculate frame interval
        frame_interval = int(video_fps / self.target_fps)
        
        frame_count = 0
        extracted_count = 0
        
        try:
            while True:
                ret, frame = cap.read()
                if not ret:
                    break
                
                # Extract every Nth frame
                if frame_count % frame_interval == 0:
                    self._send_frame(frame, frame_count, total_frames)
                    extracted_count += 1
                
                frame_count += 1
                
        finally:
            cap.release()
            print(f"\n✅ Extraction complete!")
            print(f"   Total frames: {frame_count}")
            print(f"   Extracted: {extracted_count}")
            print(f"   Sent: {self.frames_sent}")
            print(f"   Cached: {self.frames_cached}")
            print(f"   Skipped: {self.frames_skipped}")
    
    def extract_from_youtube(self, youtube_url: str, max_duration: int = 300) -> None:
        """
        Extract frames from a YouTube video
        
        Args:
            youtube_url: YouTube video URL
            max_duration: Maximum seconds to process (default: 300 = 5 minutes)
        """
        try:
            import yt_dlp
        except ImportError:
            raise ImportError("yt-dlp not installed. Run: pip install yt-dlp")
        
        print(f"📺 Downloading YouTube video: {youtube_url}")
        
        # Use temp directory that works on both Windows and Unix
        temp_dir = os.path.join(os.path.dirname(__file__), 'temp_downloads')
        os.makedirs(temp_dir, exist_ok=True)
        
        ydl_opts = {
            'format': 'worst[ext=mp4]/worst/best',  # Use lowest quality MP4 to save bandwidth
            'outtmpl': os.path.join(temp_dir, 'temp_video.%(ext)s'),
            'quiet': False,
            'no_warnings': False,
            # Try to avoid formats that require ffmpeg
            'prefer_free_formats': True,
        }
        
        try:
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(youtube_url, download=True)
                video_path = ydl.prepare_filename(info)
            
            print(f"✅ Downloaded to: {video_path}")
            
            # Extract frames from downloaded video
            self.extract_from_video(video_path)
            
            # Cleanup
            if os.path.exists(video_path):
                os.remove(video_path)
                print("🗑️  Cleaned up temporary video file")
            
            # Remove temp directory if empty
            try:
                os.rmdir(temp_dir)
            except:
                pass  # Directory not empty or doesn't exist
                
        except Exception as e:
            error_msg = str(e)
            if 'ffmpeg' in error_msg.lower():
                print("\n❌ Error: ffmpeg is required to download YouTube videos")
                print("\n📦 Install ffmpeg:")
                print("   Windows: choco install ffmpeg")
                print("   macOS:   brew install ffmpeg")
                print("   Linux:   apt install ffmpeg")
                print("\nOr download manually: https://ffmpeg.org/download.html")
                print("\n💡 Alternative: Download the video manually and use --video instead:")
                print(f"   python {__file__} --video path/to/video.mp4 --camera-id ... --api-url ...")
            raise
    
    def _send_frame(self, frame, frame_number: int, total_frames: int) -> None:
        """Send a single frame to the API"""
        # Encode frame to JPEG
        _, buffer = cv2.imencode('.jpg', frame, [cv2.IMWRITE_JPEG_QUALITY, 85])
        
        # Convert to base64
        frame_b64 = base64.b64encode(buffer).decode('utf-8')
        
        # Send to API
        payload = {
            'cameraId': self.camera_id,
            'frameData': frame_b64,
            'priority': self.priority
        }
        
        try:
            response = requests.post(self.api_url, json=payload, timeout=30)
            response.raise_for_status()
            result = response.json()
            
            # Track results
            if result.get('cached'):
                self.frames_cached += 1
                status = "💾 CACHED"
            elif result.get('skipped'):
                self.frames_skipped += 1
                status = "⏭️  SKIPPED"
            elif result.get('queued'):
                self.frames_sent += 1
                status = f"📤 QUEUED {result.get('message', '')}"
            else:
                status = "❓ UNKNOWN"
            
            progress = (frame_number / total_frames) * 100
            print(f"[{progress:5.1f}%] Frame {frame_number:6d}: {status}")
            
        except requests.exceptions.RequestException as e:
            print(f"❌ Error sending frame {frame_number}: {e}")
    
    def extract_from_webcam(self, duration: int = 60) -> None:
        """
        Extract frames from webcam for live testing
        
        Args:
            duration: Duration in seconds to capture (default: 60)
        """
        print(f"📷 Opening webcam (will capture for {duration}s)")
        cap = cv2.VideoCapture(0)
        
        if not cap.isOpened():
            raise ValueError("Failed to open webcam")
        
        start_time = time.time()
        frame_count = 0
        
        try:
            while (time.time() - start_time) < duration:
                ret, frame = cap.read()
                if not ret:
                    break
                
                self._send_frame(frame, frame_count, duration * self.target_fps)
                frame_count += 1
                
                # Wait for next frame based on target FPS
                time.sleep(1.0 / self.target_fps)
                
        finally:
            cap.release()
            print(f"\n✅ Webcam capture complete!")
            print(f"   Frames captured: {frame_count}")
            print(f"   Sent: {self.frames_sent}")
            print(f"   Cached: {self.frames_cached}")


def main():
    parser = argparse.ArgumentParser(
        description="Extract frames from video and send to Convex for analysis"
    )
    
    # Input source (one required)
    source_group = parser.add_mutually_exclusive_group(required=True)
    source_group.add_argument('--video', help='Path to video file')
    source_group.add_argument('--youtube', help='YouTube video URL')
    source_group.add_argument('--webcam', action='store_true', help='Use webcam')
    
    # Required parameters
    parser.add_argument('--camera-id', required=True, help='Convex camera feed ID')
    parser.add_argument('--api-url', required=True, help='Convex deployment URL')
    
    # Optional parameters
    parser.add_argument('--fps', type=float, default=0.5, 
                       help='Target FPS for extraction (default: 0.5 = 1 frame per 2 seconds)')
    parser.add_argument('--priority', type=int, default=1, choices=range(1, 11),
                       help='Frame priority 1-10 (default: 1)')
    parser.add_argument('--duration', type=int, default=60,
                       help='Duration in seconds for webcam capture (default: 60)')
    
    args = parser.parse_args()
    
    # Initialize extractor
    extractor = FrameExtractor(
        api_url=args.api_url,
        camera_id=args.camera_id,
        target_fps=args.fps,
        priority=args.priority
    )
    
    # Extract from appropriate source
    if args.video:
        extractor.extract_from_video(args.video)
    elif args.youtube:
        extractor.extract_from_youtube(args.youtube)
    elif args.webcam:
        extractor.extract_from_webcam(args.duration)


if __name__ == '__main__':
    main()

